### PR TITLE
BUG: Fix regression for groupby.indices in case of unused categories

### LINKED ISSUE #38642: REGR: GroupBy.indices no longer includes unobserved categories
Does anybody know if this was an intentional change? (I don't directly find something about it in the whatsnew)

```
In [9]: pd.__version__ 
Out[9]: '1.0.5'

In [10]: df = pd.DataFrame({"key": pd.Categorical(["b"]*5, categories=["a", "b", "c", "d"]), "col": range(5)}) 

In [11]: gb = df.groupby("key")   

In [12]: list(gb.indices)  
Out[12]: ['a', 'b', 'c', 'd']
```

vs

```
In [1]: pd.__version__
Out[1]: '1.3.0.dev0+92.ga2d10ba88a'

In [2]: df = pd.DataFrame({"key": pd.Categorical(["b"]*5, categories=["a", "b", "c", "d"]), "col": range(5)})

In [3]: gb = df.groupby("key")

In [4]: list(gb.indices)
Out[4]: ['b']
```

This already changed in pandas 1.1, so not a recent change.

The consequence of this is that iterating over `gb` vs iterating over `gb.indices` is not consistent anymore.

cc @mroeschke @rhshadrach

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #38642
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

 To show the unused we categories, we can dispatch back as we did before #36842. Since Categorical does not support nan, the reason for the switch does not exist in this case. I think handling unused categories in ``get_indexer_dict`` is not desirable, because it would introduce a lot of convolution for one corner case we could handle here pretty easily.
Since we are only using ``indices`` in grouper only for categoricals, we do not have to worry about sorting as the test shows.

cc @jorisvandenbossche @mroeschke

### CODE REVIEW

File: pandas/core/groupby/ops.py [hunk 1]
```diff
@@ -241,6 +242,11 @@ def apply(self, f: F, data: FrameOrSeries, axis: int = 0):
     @cache_readonly
     def indices(self):
         """ dict {group name -> group indices} """
+        if len(self.groupings) == 1 and isinstance(
```
Comment (jreback) [inline pandas/core/groupby/ops.py:245]: woa, this is so special casing here. why is this added?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/ops.py (hunk 1 above)
Comment (jorisvandenbossche) [reply pandas/core/groupby/ops.py:245]: See the explanation in the top post

**Comment by jorisvandenbossche** [comment]
> Since Categorical does not support nan, the reason for the switch does not exist in this case.

@phofl Categorical does actually support NaNs, but maybe this is not covered by the tests?

**Comment by jorisvandenbossche** [comment]
It seems this might actually also be broken for Categorical (or not fixed by https://github.com/pandas-dev/pandas/pull/36842). This is what I see on master:

```
In [8]: df = pd.DataFrame({'key': pd.Categorical(['a', 'b', 'a', 'b', np.nan]), 'col': range(5)})

In [9]: df.groupby('key', dropna=False).indices
Out[9]: {'b': array([1, 3]), 'a': array([0, 2])}

In [10]: df.astype({"key": object}).groupby('key', dropna=False).indices
Out[10]: {'a': array([0, 2]), 'b': array([1, 3]), nan: array([4])}
```

File: pandas/core/groupby/ops.py (hunk 1 above)
Comment (phofl) [reply pandas/core/groupby/ops.py:245]: thx

**Comment by phofl** [comment]
Hm, I misunderstood https://github.com/pandas-dev/pandas/issues/35646#issuecomment-687879101 this then. Have to look into this then

File: pandas/core/groupby/ops.py (hunk 1 above)
Comment (jreback) [reply pandas/core/groupby/ops.py:245]: I read it and don't buy it. Why are we supporting this case at all? I would rather simply fix this breaking or not.

**Comment by jorisvandenbossche** [comment]
Ah, no, I was not aware of that categorical with ``dropna=False`` might be knowingly not supported at the moment. 
cc @mroeschke is there a reason that we wouldn't want to support that?

File: pandas/core/groupby/ops.py (hunk 1 above)
Comment (phofl) [reply pandas/core/groupby/ops.py:245]: You mean not showing unused categories for indices?

Comment (jreback) [reply pandas/core/groupby/ops.py:245]: i c we did thave this special case before, but *onl* was for len(self.groupins) == 1. its the special isinstance i object.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by mroeschke** [comment]
My comment was assuming that that adding a new `dropna` keyword argument to `Categorical` indicating that `nan` shouldn't be a missing value but a category itself wouldn't be a popular change. That assumption might be incorrect though

**Comment by jreback** [comment]
> My comment was assuming that that adding a new `dropna` keyword argument to `Categorical` indicating that `nan` shouldn't be a missing value but a category itself wouldn't be a popular change. That assumption might be incorrect though

not happening, we already removed this case thru a deprecation cycle and no reason to add it back

**Comment by jorisvandenbossche** [comment]
Categorical indeed does not support NaN *as a category*, but does that mean it could not be supported in `groupby(dropna=False)`? A categorical can have missing *values*, so the resulting index of `groupby(dropna=False)` would not have NaN as category, but as missing value (as -1 in the codes)

**Comment by mroeschke** [comment]
I imagine it should be supported in `groupby(dropna=False)`.

At the time I was probably focusing too much on `Grouper.indices` unaware that there was also `BaseGrouper.indices` (and the relationship between them) because `Grouper.indices` was relevant for a groupby rolling bug I was hunting down.

**Comment by jreback** [comment]
@phofl can you merge master and add a note for 1.2.1

**Comment by phofl** [comment]
Done

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @phofl very nice

**Comment by jreback** [comment]
@meeseeksdev backport 1.2.x
